File: andromeda/src/models/roleModels.ts
import { Validation, GetOptions } from "./commonModels";
import { DepartmentType } from "./departmentModels";

export interface Role {
    id?: number;
    name: string;
    canTeach: boolean;
    minLoad?: number;
    maxLoad?: number;

    roleDepartments: RoleInDepartment[];
}

export namespace Role {
    export const initial: Role = {
        name: '',
        roleDepartments: [],
        canTeach: false,
        minLoad: null,
        maxLoad: null
    }
}

export interface RoleInDepartment {
    id?: number;
    roleId?: number;
    departmentId?: number;
    departmentType?: DepartmentType;
    roleName?: string;
    departmentName?: string;
}

export interface RoleGetOptions extends GetOptions {

}

export interface RoleValidation extends Validation {
    nameError?: string;
    minLoadError?: string;
    maxLoadError?: string;
}

export namespace RoleValidation {
    export const initial: RoleValidation = Validation.initial;
}

File: Andromeda.Data/Models/RoleModels.cs
using System.Collections.Generic;
using Andromeda.Data.Enumerations;

namespace Andromeda.Data.Models
{
    public class Role
    {
        public int Id { get; set; }
        public string Name { get; set; }

        public List<RoleInDepartment> RoleDepartments { get; set; }
    }

    public class RoleInDepartment
    {
        public int Id { get; set; }
        public int RoleId { get; set; }
        public int DepartmentId { get; set; }
        public DepartmentType DepartmentType { get; set; }
        public string RoleName { get; set; }
        public string DepartmentName { get; set; }
    }

    public class UserRoleInDepartment
    {
        public int Id { get; set; }
        public int RoleInDepartmentId { get; set; }
        public int RoleId { get; set; }
        public int DepartmentId { get; set; }
        public int UserId { get; set; }

        public string RoleName { get; set; }
        public string DepartmentName { get; set; }
        public string UserFullName { get; set; }
    }

    public class RoleGetOptions : BaseGetOptions
    {
        public string NormalizedSearch => !string.IsNullOrEmpty(Search) ? $"%{Search}%" : string.Empty;
        public string Search { get; set; }
        public string Name { get; set; }
    }

    public class RoleInDepartmentGetOptions
    {
        public int? RoleId { get; set; }
        public IReadOnlyList<int> RoleIds { get; set; }
        public IReadOnlyList<int> DepartmentIds { get; set; }
        public int? DeparmtentId { get; set; }
    }

    public class UserRoleInDepartmentGetOptions
    {
        public int? DepartmentId { get; set; }

        public IReadOnlyList<int> RoleIds { get; set; }
        public IReadOnlyList<int> UserIds { get; set; }
        public IReadOnlyList<int> DepartmentIds { get; set; }
    }
}